frontend/model/notifications/vuexModule.js

```diff
@@ -0,0 +1,104 @@
+import type {
+  Notification,
+  VuexModuleContext
+} from './types.flow.js'
+
+import sbp from '~/shared/sbp.js'
+import './selectors.js'
+import { age, isNew, isOlder } from './utils.js'
+import * as keys from './mutationKeys.js'
+
+const actions = {
+  async login (context: VuexModuleContext, user: Object) {
+    const notifications = await sbp('gi.db/notifications/load', user.username)
+
+    if (notifications && notifications.length) {
+      context.commit(keys.SET_NOTIFICATIONS, notifications)
+    }
+  },
+  async logout (context: VuexModuleContext) {
+    if (context.rootState.loggedIn) {
+      // Make sure to save the notifications *before* clearing them.
+      await sbp('gi.notifications/save')
+      context.commit(keys.REMOVE_ALL_NOTIFICATIONS)
+    }
+  }
+}
+
+const getters = {
+  newNotifications (state, getters) {
+    return state.filter(item => isNew(item))
+  },
+
+  newNotificationCount (state, getters) {
+    return getters.newNotifications.length
+  },
+
+  notificationCount (state, getters) {
+    return state.length
+  },
+
+  olderNotifications (state, getters) {
+    return state.filter(item => isOlder(item))
+  },
+
+  olderNotificationCount (state, getters) {
+    return getters.olderNotifications.length
+  },
+
+  unreadNotifications (state, getters) {
+    return state.filter(item => !item.read)
+  },
+
+  unreadNotificationCount (state, getters) {
+    return getters.unreadNotifications.length
+  }
+}
+
+const mutations = {
+  [keys.ADD_NOTIFICATION] (state, notification: Notification) {
+    if (state.includes(notification)) {
+      throw new Error('This notification is already in the store.')
+    }
+    if (state[0] && age(notification) > age(state[0])) {
+      throw new Error('This notification is older than the latest one in the store.')
+    }
+    // Using `unshift()` here to insert the new item before older ones.
+    state.unshift(notification)
+  },
+
+  [keys.MARK_NOTIFICATION_AS_READ] (state, notification: Notification) {
+    notification.read = true
+  },
+
+  [keys.MARK_ALL_NOTIFICATIONS_AS_READ] (state) {
+    state.forEach(item => {
+      item.read = true
+    })
+  },
+
+  [keys.REMOVE_NOTIFICATION] (state, notification: Notification) {
+    const index = state.indexOf(notification)
+
+    if (~index) {
```

Comment:
Nothing wrong here, but it seems a bit of an odd way (as in, I'm unaccustomed to see bitwise operations in JS) to check. I'd _subjectively_ say `index !== -1` or `index >= 0` are more idiomatic.

Reply:
Yes, this idiom is less common than its counterparts.

Which I subjectively think is a bit unfortunate, since it is arguably less visual clutter, but also avoids having to consistently choose between` index !== -1`, `index >= 0`, but also `index > -1`, which is nice because none of them is clearly better than the other two.

So I've used this idiom here, waiting for feedback, but I will replace it by another it required.

Reply:
I've seen people do this but I agree with @corrideat personally... because it's not clear to a human what the value of `~index` is. By bounding it with `index > -1` or `index >= 0` it's clear what the intent behind the `if` is. So if it were me I'd change it, but @snowteamer if you feel strongly that you want to keep it we can.

Reply:
That's no big deal, we can change it. Which variant would you prefer?

Reply:
Going with `index > -1` for now.